### ISSUE TITLE
Matrix size incompatible error

### ISSUE THREAD

**Issue by amjass12** [danielegrattarola/spektral#207]
Hi! 

I am trying to build a basic GCN with spektral 1.0 and am attempting to do some node classification. I am getting an error as a result of the input shape. My code is below and the error underneath.

I am feeding the adjacency matrix for the graph as well as for the node features. For the graph, the adjacency matrix is pre-processed using `GCNConv.preprocess` and the node features are the adjacency matrix but with 0's and 1's to represent node degree.

Please note for A_in, sparse is set to False. The code is below:

```
adj.shape = (6541,6541)

A = GCNConv.preprocess(adj.values)
A = A.astype('f4')
A.shape=(6541,6541)

N = 6541
F = 6541
n_classes = trainingtarget.shape[1]

X_in=Input(shape=(F,))
A_in = Input(shape=(N,), sparse=False)

X_1 = GCNConv(128, 'relu')([X_in, A_in])
X_1 = tf.keras.layers.Dropout(0.2)(X_1)
X_2 = GCNConv(103, activation='softmax')([X_1, A_in])#103 communities to classify nodes for

model = Model(inputs=[X_in, A_in], outputs=X_2)

model.compile(optimizer='adam', loss='categorical_crossentropy', weighted_metrics=['acc'])

model.fit([adj.values, A], trainingtarget, shuffle=False, epochs=100)
```
The error I get from running this code is:
InvalidArgumentError:  Matrix size-incompatible: In[0]: [32,6541], In[1]: [32,128]
Errors may have originated from an input operation.
Input Source operations connected to node model_18/gcn_conv_41/BiasAdd:
 model_18/gcn_conv_41/MatMul_1 - **gcn_conv_41 is the first hidden layer (128, 'relu')**

also note that when I execute the first GCNConv layer:

```
X_1 = GCNConv(128, 'relu')([X_in, A_in])
```
WARNING:tensorflow:AutoGraph could not transform <function GCNConv.call at 0x7f7ed8d8d950> and will run it as-is.

and finally - if i set `sparse =True`
I get the following error - 

ValueError: The two structures don't have the same nested structure.

First structure: type=TensorSpec str=TensorSpec(shape=(6541, 6541), dtype=tf.float32, name=None)

Second structure: type=SparseTensor str=SparseTensor(indices=Tensor("input_53/indices:0", shape=(None, 2), dtype=int64), values=Tensor("input_53/values:0", shape=(None,), dtype=float32), dense_shape=Tensor("input_53/shape:0", shape=(2,), dtype=int64))

I know there is abit to unpack here, but i get multiple warnings/errors along the way and am wondering if there is something obvious that I am missing here. Any help moving forward is much appreciated!

thank you!

**Comment by danielegrattarola** [comment]
Hi, 

a few comments that might help you solve the issue. 

This line: 

```py
model.fit([adj.values, A], trainingtarget, shuffle=False, epochs=100)
```

automatically sets the batch size to 32, but GNNs require that all the graph is given as input at the same time. Try setting `batch_size=N` and the first issue should be solved.

About the second problem, the issue is that when you set `sparse=True` the model will expect the input to be a SparseTensor (and the conversion does not happen automatically in Keras). 

You can either convert it manually: 

```py
from spektra.layers import ops
A = ops.sp_matrix_to_sp_tensor(A)  # this will work also for numpy arrays
```

or switch to using `spektral.data.SingleLoader` for dealing with training. This last option will also solve the first problem that you have. 

Have a look at the examples: https://github.com/danielegrattarola/spektral/blob/master/examples/node_prediction/citation_gcn.py

Cheers

**Comment by amjass12** [comment]
Hi @danielegrattarola , 

thank you for your detailed reply - I can confirm this all works now batch_size = 6549 and this now starts training (if I am correct, each individual node has to be treated as a separate 'sample' right?)

I can also confirm the sparse tensor approach works - however i find that training is slightly slower than when `sparse=False`.. is there a reason for this?

Also - the warning highlighted above `WARNING:tensorflow:AutoGraph could not transform` was caused by a different version of gast. I downgraded gast to 0.2 and this warning no longer shows. 

thank you!
